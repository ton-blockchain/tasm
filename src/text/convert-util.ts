import * as $ from "./util"
import type {$ast} from "./grammar"
import type {Loc} from "../runtime/util"

export const assertZeroArgs = (instr: $ast.Instruction, loc: Loc) => {
    const argsLen = instr.args
    if (argsLen.length > 0) {
        throw new $.ParseError(loc, "Expected 0 arguments")
    }
}

export const assertSingleArgs = (instr: $ast.Instruction, loc: Loc) => {
    const argsLen = instr.args
    if (argsLen.length !== 1) {
        throw new $.ParseError(loc, "Expected 1 argument")
    }
}
