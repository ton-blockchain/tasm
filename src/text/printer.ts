import type {Instr} from "../runtime"
import type {Code, Dict} from "../runtime/util"
import type {Slice} from "@ton/core"
import {printInstruction} from "./printer-gen"

export class Printer {
    public indent: number
    public result: string

    public constructor(indent: number) {
        this.indent = indent
        this.result = ""
    }

    public beginLine(text: string) {
        const indentText = "    ".repeat(this.indent)
        this.result += indentText + text
    }

    public append(text: string) {
        this.result += text
    }

    public inIndent(action: (p: Printer) => void) {
        this.indent++
        action(this)
        this.indent--
    }

    public build(): string {
        return this.result
    }
}

export const printInstr = (p: Printer, instr: Instr) => {
    if (instr.$ === "PSEUDO_PUSHREF") {
        p.beginLine("ref ")
        printCode(p, instr.arg0)
        return
    }

    if (instr.$ === "PSEUDO_PUSHSLICE") {
        p.beginLine("embed ")
        printSlice(p, instr.arg0)
        return
    }

    if (instr.$ === "PSEUDO_EXOTIC") {
        p.beginLine("exotic ")

        if (instr.arg0.$ === "LibraryCell") {
            p.append("library ")
            printSlice(p, instr.arg0.data)
            return
        }

        printSlice(p, instr.arg0.cell.beginParse(true))
        return
    }

    printInstruction(p, instr)
}

export const printInstructions = (p: Printer, instructions: Instr[]) => {
    if (instructions.length === 0) {
        p.append("{}")
        return
    }

    p.append("{\n")
    p.inIndent(p => {
        for (const instr of instructions) {
            printInstr(p, instr)
            p.append("\n")
        }
    })
    p.beginLine("}")
}

export const print = (instructions: Instr[]) => {
    const p = new Printer(0)
    for (const instr of instructions) {
        printInstr(p, instr)
        p.append("\n")
    }
    return p.build()
}

export function printDictionary(p: Printer, dict: Dict) {
    if (dict.$ === "RawDict") {
        printSlice(p, dict.slice)
    }

    if (dict.$ === "DecompiledDict") {
        p.append(`[\n`)

        for (const method of dict.methods) {
            p.inIndent(p => {
                p.beginLine(`${method.id} => `)
                printInstructions(p, method.instructions)
                p.append(`\n`)
            })
        }

        p.beginLine("]")
    }
}

export function printCode(p: Printer, arg0: Code) {
    if (arg0.$ === "Instructions") {
        printInstructions(p, arg0.instructions)
    }

    if (arg0.$ === "Raw") {
        printSlice(p, arg0.slice)
    }
}

export function printSlice(p: Printer, slice: Slice) {
    const hasRefs = slice.remainingRefs
    if (hasRefs) {
        const data = slice.asCell().toBoc().toString("hex")
        p.append(`boc{${data}}`)
    } else {
        const data = slice.asCell().bits.toString()
        p.append(`x{${data}}`)
    }
}
