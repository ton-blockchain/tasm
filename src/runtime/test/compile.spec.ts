import {compileCellWithMapping, decompileCell} from "../instr"
import {parse, print} from "../../text"
import * as fs from "node:fs"
import {diffLines} from "diff"

const test = (text: string): (() => void) => {
    return () => {
        const res = parse("test.asm", text)
        if (res.$ === "ParseFailure") {
            return
        }

        const [compiled] = compileCellWithMapping(res.instructions, {skipRefs: true})
        expect(compiled.hash().toString("hex")).toMatchSnapshot()

        const withoutDebugMarks = text
            .split(/\n/)
            .filter(line => !line.includes("DEBUGMARK"))
            .join("\n")

        const res2 = parse("test.asm", withoutDebugMarks)
        if (res2.$ === "ParseFailure") {
            return
        }

        const [compiled2] = compileCellWithMapping(res2.instructions, {skipRefs: true})
        expect(compiled2.hash().toString("hex")).toMatchSnapshot()

        expect(compiled.hash().toString("hex")).toEqual(compiled2.hash().toString("hex"))

        const compiledAssembly = print(decompileCell(compiled))
        const compiledAssembly2 = print(decompileCell(compiled2))

        if (compiledAssembly !== compiledAssembly2) {
            console.log(diffLines(compiledAssembly, compiledAssembly2))
            expect(true).toBe(false)
        }
    }
}

describe("Compile", () => {
    it(
        "should compile Vault with skipRefs option enabled",
        test(fs.readFileSync(`${__dirname}/testdata/Vault.tasm`, "utf8")),
    )
})
