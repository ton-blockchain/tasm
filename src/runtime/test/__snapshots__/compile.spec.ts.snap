// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile should compile Vault with skipRefs option enabled 1`] = `"606d59fd3f7551db56c690541e13026a47ed5df8c2c945177aeca543a8449027"`;

exports[`Compile should compile Vault with skipRefs option enabled 2`] = `"606d59fd3f7551db56c690541e13026a47ed5df8c2c945177aeca543a8449027"`;
